bookkeeper-server/src/test/java/org/apache/bookkeeper/client/api/BookKeeperBuildersTest.java

```diff
@@ -0,0 +1,458 @@
+/*
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+package org.apache.bookkeeper.client.api;
+
+import io.netty.buffer.Unpooled;
+import java.lang.reflect.Method;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import org.apache.bookkeeper.client.BKException;
+import org.apache.bookkeeper.client.BKException.BKDigestMatchException;
+import org.apache.bookkeeper.client.BKException.BKNoSuchLedgerExistsException;
+import org.apache.bookkeeper.client.BKException.BKUnauthorizedAccessException;
+import org.apache.bookkeeper.client.LedgerEntry;
+import org.apache.bookkeeper.client.LedgerHandle;
+import org.apache.bookkeeper.client.LedgerMetadata;
+import org.apache.bookkeeper.client.MacDigestManager;
+import org.apache.bookkeeper.common.concurrent.FutureUtils;
+import org.apache.bookkeeper.conf.ClientConfiguration;
+import org.apache.bookkeeper.test.BookKeeperClusterTestCase;
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+import org.junit.Test;
+
+/**
+ * Unit tests of builders in this package
+ */
+public class BookKeeperBuildersTest extends BookKeeperClusterTestCase {
```

Comment:
please test builder using mockito. it doesn't make any sense to start a cluster of bookie just for testing builders. the current test suite is already taking very long time to run. we should avoid run bookies if necessary.

Reply:
I think that we are testing the behavior not only of the builders but of the new API.
Builders are implemented with special behavior of the Ops and we must be sure that LedgerHandles created with the new API really do what they are expected to do.

Anyway I would create another test file with mocks only to test the fields

Reply:
I strongly feel it doesn't need to do such test, because it unnecessarily increase the time to run the test suite. Because this change doesn't actually change any behavior, it just changes the way how do we construct and build the client and operations. So we should just test the builder behaviors only.

Reply:
I am paranoid about the tests we are adding to bookkeeper, because of the way how we did test, current test suite has quite a few flaky tests. I would like us to think about what should be really tested before spawning a cluster of N bookies. It is most likely to introduce flakiness into the test suite.